sandbox/apparmor/profile.go

```diff
@@ -251,3 +302,101 @@ var SnapConfineDistroProfilePath = func() string {
+		return false, fmt.Errorf("cannot create snap-confine policy directory: %s", err)
+	}
+
+	// Location of the generated policy.
+	glob := "*"
+	policy := make(map[string]osutil.FileState)
+
+	// Check if NFS is mounted at or under $HOME. Because NFS is not
+	// transparent to apparmor we must alter our profile to counter that and
+	// allow snap-confine to work.
+	if nfs, err := osutilIsHomeUsingNFS(); err != nil {
+		logger.Noticef("cannot determine if NFS is in use: %v", err)
+	} else if nfs {
+		policy["nfs-support"] = &osutil.MemoryFileState{
+			Content: []byte(NfsSnippet),
+			Mode:    0644,
+		}
+		logger.Noticef("snapd enabled NFS support, additional implicit network permissions granted")
+	}
+
+	// Check if '/' is on overlayfs. If so, add the necessary rules for
+	// upperdir and allow snap-confine to work.
+	if overlayRoot, err := osutilIsRootWritableOverlay(); err != nil {
+		logger.Noticef("cannot determine if root filesystem on overlay: %v", err)
+	} else if overlayRoot != "" {
+		snippet := strings.Replace(OverlayRootSnippet, "###UPPERDIR###", overlayRoot, -1)
+		policy["overlay-root"] = &osutil.MemoryFileState{
+			Content: []byte(snippet),
+			Mode:    0644,
+		}
+		logger.Noticef("snapd enabled root filesystem on overlay support, additional upperdir permissions granted")
+	}
+
+	// Check whether apparmor_parser supports bpf capability. Some older
+	// versions do not, hence the capability cannot be part of the default
+	// profile of snap-confine as loading it would fail.
+	if features, err := ParserFeatures(); err != nil {
+		logger.Noticef("cannot determine apparmor_parser features: %v", err)
+	} else if strutil.ListContains(features, "cap-bpf") {
+		policy["cap-bpf"] = &osutil.MemoryFileState{
+			Content: []byte(capabilityBPFSnippet),
+			Mode:    0644,
+		}
+	}
+
+	if len(homedirs) > 0 {
+		policy["homedirs"] = &osutil.MemoryFileState{
+			Content: []byte(snapConfineHomedirsSnippet(homedirs)),
+			Mode:    0644,
+		}
+	}
+
+	// Ensure that generated policy is what we computed above.
+	created, removed, err := osutil.EnsureDirState(dirs.SnapConfineAppArmorDir, glob, policy)
+	if err != nil {
+		return false, fmt.Errorf("cannot synchronize snap-confine policy: %s", err)
+	}
+	return len(created) > 0 || len(removed) > 0, nil
+}
+
+func RemoveSnapConfineSnippets() error {
+	glob := "*"
+	_, _, err := osutil.EnsureDirState(dirs.SnapConfineAppArmorDir, glob, nil)
+	return err
+}
+
+// Reload the profile of snap-confine. This function must be called when the
+// AppArmor profile snippets for snap-confine have been changed. Note that
+// there might be more than one of them in the system, because besides the one
+// coming with the system installation, also the snapd and core snaps ship a
+// snap-confine binary. Of course, only one of these binaries will be used by
+// snapd, but in this context we don't want to deal with this logic, so let's
+// just reload all of them.
+var ReloadSnapConfineProfile = func() error {
```

Comment:
I'm confused, I don't see calls to this?

Reply:
It seems to be a unfinished part of refactoring some code from backend to profile. I'll remove this right now and do this in a followup, I don't believe there is any need for this to be a part of this PR